File: ios-renderer/Sources/A2UIRenderer/Processing/DataBindingResolver.swift
import Foundation

/// Resolves data binding references like "${/hello_world/text}" → "Hello world".
/// Port of resolveDataBinding() from a2ui-parser/src/index.ts.
public struct DataBindingResolver {
    public let dataModel: [String: JSONValue]

    public init(dataModel: [String: JSONValue]) {
        self.dataModel = dataModel
    }

    /// Resolve a data binding reference.
    /// - "${/hello_world_our_1st_tile/text}" → "Hello world, our 1st tile!"
    /// - "Static text" → "Static text" (passthrough)
    /// - nil → ""
    public func resolve(_ value: String?) -> String {
        guard let value, !value.isEmpty else { return "" }

        // Check for data binding pattern: ${/path/to/value}
        guard value.hasPrefix("${/"), value.hasSuffix("}") else {
            return resolveTemplateTokens(value)
        }

        let path = String(value.dropFirst(3).dropLast()) // strip "${/" and "}"
        let parts = path.split(separator: "/").map(String.init)

        var current: JSONValue = .object(dataModel)
        for part in parts {
            guard let obj = current.objectValue, let next = obj[part] else {
                return resolveTemplateTokens(value) // Return original if path doesn't resolve
            }
            current = next
        }

        switch current {
        case .string(let s): return resolveTemplateTokens(s)
        case .number(let n):
            if n == n.rounded() && n < 1_000_000 {
                return String(Int(n))
            }
            return String(n)
        case .bool(let b): return String(b)
        default: return resolveTemplateTokens(value)
        }
    }

    /// Replace template tokens like {{current_time}}, {{current_date}} with device values.
    private func resolveTemplateTokens(_ text: String) -> String {
        guard text.contains("{{") else { return text }

        var result = text
        let now = Date()

        if result.contains("{{current_time}}") {
            let formatter = DateFormatter()
            formatter.dateFormat = "h:mm a"
            result = result.replacingOccurrences(of: "{{current_time}}", with: formatter.string(from: now))
        }

        if result.contains("{{current_time_24h}}") {
            let formatter = DateFormatter()
            formatter.dateFormat = "HH:mm"
            result = result.replacingOccurrences(of: "{{current_time_24h}}", with: formatter.string(from: now))
        }

        if result.contains("{{current_date}}") {
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            result = result.replacingOccurrences(of: "{{current_date}}", with: formatter.string(from: now))
        }

        if result.contains("{{current_day}}") {
            let formatter = DateFormatter()
            formatter.dateFormat = "EEEE"
            result = result.replacingOccurrences(of: "{{current_day}}", with: formatter.string(from: now))
        }

        return result
    }
}

File: android-renderer/lib/src/main/kotlin/com/interfaceai/a2ui/processing/DataBindingResolver.kt
package com.interfaceai.a2ui.processing

import kotlinx.serialization.json.*

/**
 * Resolves data binding references like "${/hello_world/text}" → "Hello world".
 * Port of resolveDataBinding() from a2ui-parser/src/index.ts.
 */
class DataBindingResolver(val dataModel: Map<String, JsonElement>) {

    /**
     * Resolve a data binding reference.
     * "${/hello_world_our_1st_tile/text}" → "Hello world, our 1st tile!"
     * "Static text" → "Static text" (passthrough)
     * null → ""
     */
    fun resolve(value: String?): String {
        if (value.isNullOrEmpty()) return ""

        // Check for data binding pattern: ${/path/to/value}
        if (!value.startsWith("\${/") || !value.endsWith("}")) return value

        val path = value.substring(3, value.length - 1) // strip "${/" and "}"
        val parts = path.split("/")

        var current: JsonElement = JsonObject(dataModel)
        for (part in parts) {
            val obj = (current as? JsonObject) ?: return value
            current = obj[part] ?: return value
        }

        return when {
            current is JsonPrimitive && current.isString -> current.content
            current is JsonPrimitive -> current.content
            else -> value
        }
    }
}
